hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactingMemStore.java

```diff
@@ -0,0 +1,565 @@
+  }
+
+  /**
+   * On flush, how much memory we will clear.
+   * @return size of data that is going to be flushed
+   */
+  @Override public long getFlushableSize() {
+    long snapshotSize = getSnapshot().getSize();
+    if(snapshotSize == 0) {
+      snapshotSize = pipeline.getTailSize();
+    }
+    return snapshotSize > 0 ? snapshotSize : keySize();
+  }
+
+  @Override
+  public void updateLowestUnflushedSequenceIdInWal(boolean onlyIfGreater) {
+    long minSequenceId = pipeline.getMinSequenceId();
+    if(minSequenceId != Long.MAX_VALUE) {
+      byte[] encodedRegionName = getRegionServices().getRegionInfo().getEncodedNameAsBytes();
+      byte[] familyName = getFamilyName();
+      WAL wal = getRegionServices().getWAL();
+      if (wal != null) {
+        wal.updateStore(encodedRegionName, familyName, minSequenceId, onlyIfGreater);
+      }
+    }
+  }
+
+  @Override
+  public LinkedList<Segment> getListOfSegments() {
+    LinkedList<Segment> pipelineList = pipeline.getStoreSegmentList();
+    LinkedList<Segment> list = new LinkedList<Segment>();
+    list.add(getActive());
+    list.addAll(pipelineList);
+    list.add(getSnapshot());
+    return list;
+  }
+
+  @Override
+  protected List<SegmentScanner> getListOfScanners(long readPt) throws IOException {
+    LinkedList<Segment> pipelineList = pipeline.getStoreSegmentList();
+    LinkedList<SegmentScanner> list = new LinkedList<SegmentScanner>();
+    list.add(getActive().getSegmentScanner(readPt));
+    for (Segment item : pipelineList) {
+      list.add(item.getSegmentScanner(readPt));
+    }
+    list.add(getSnapshot().getSegmentScanner
+        (readPt));
+    // set sequence ids by decsending order
+    Iterator<SegmentScanner> iterator = list.descendingIterator();
+    int seqId = 0;
+    while (iterator.hasNext()) {
+      iterator.next().setSequenceID(seqId);
+      seqId++;
+    }
+    return list;
+  }
+
+  /**
+   * Check whether anything need to be done based on the current active set size.
+   * The method is invoked on every addition to the active set.
+   * For CompactingMemStore, flush the active set to the read-only memory if it's
+   * size is above threshold
+   */
+  @Override
+  protected void checkActiveSize() {
+    if (shouldFlushInMemory()) {
+      /* The thread is dispatched to flush-in-memory. This cannot be done
+      * on the same thread, because for flush-in-memory we require updatesLock
+      * in exclusive mode while this method (checkActiveSize) is invoked holding updatesLock
+      * in the shared mode. */
+      ExecutorService pool = getPool();
+      if(pool != null) {
+        InMemoryFlushWorker worker = new InMemoryFlushWorker();
+        LOG.info("Dispatching the MemStore in-memory flush for store "
+            + store.getColumnFamilyName());
+        pool.submit(worker);
+        inMemoryFlushInProgress.set(true);
+      }
+    }
+  }
+
+  // internal method, visible only for tests
```

Comment:
Suggestion to improve the comment: "internally used method, externally visible only for tests"